application_manager/plugins/spark_generic/plugin.py

```diff
@@ -0,0 +1,359 @@
+                              {'dependencies': dependencies,
+                               'main_class': main_class,
+                               'master': remote_instance,
+                               'job_binary_file': 'file://'+job_binary_file,
+                               'args': args_line})
+
+        if main_class == '':
+            spark_submit = spark_submit.replace('--class', '')
+
+        if dependencies == '':
+            spark_submit = spark_submit.replace('--packages', '')
+
+        job = remote.execute_command_popen(remote_instance, 
+                                           key_path, 
+                                           spark_submit)
+
+        return job
+
+    def _enable_event_log(self, master, key_path, path):
+        enable_event_log_command = ("echo -e 'spark.executor.extraClassPath "
+             "/usr/lib/hadoop-mapreduce/hadoop-openstack.jar\n"
+             "spark.eventLog.enabled true\n"
+             "spark.eventLog.dir "
+             "file://%(path)s' > "
+             "/opt/spark/conf/spark-defaults.conf" % {'path': path})
+
+        remote.execute_command(master, key_path, enable_event_log_command)
+
+    def _log(self, string):
+        plugin_log.log(string)
+        self.running_log.log(string)
+
+    def _verify_existing_log_paths(self, app_id):
+        if not os.path.exists('logs'):
+            os.mkdir('logs')
+        elif not os.path.exists('logs/apps'):
+            os.mkdir('logs/apps')
+        if not os.path.exists('logs/apps/%s' % app_id):
+            os.mkdir('logs/apps/%s' % app_id)
+
+    def _clean_log_files(self, app_id):
+        running_log_file = open("logs/apps/%s/execution" % app_id, "w").close()
+        stdout_file = open("logs/apps/%s/stdout" % app_id, "w").close()
+        stderr_file = open("logs/apps/%s/stderr" % app_id, "w").close()
+
+    def _mkdir(self, path):
+        subprocess.call('mkdir -p %s' % path, shell=True) 
+
+ 
+class SparkGenericProvider(base.PluginInterface):
+
+    def __init__(self):
+        self.applications = []
+        self.spark_applications_ids = []
+        self.id_generator = ID_Generator()
+        self.masters = {ip : False for ip in api.masters_ips}
+
+    def get_title(self):
+        return 'Spark Generic'
+
+    def get_description(self):
+        return 'Plugin that allows utilization of Spark clusters to run jobs'
+
+    def to_dict(self):
+        return {
+            'name': self.name,
+            'title': self.get_title(),
+            'description': self.get_description(),
+        }
+
+    def update_running_clusters(self):
+        for application in self.applications:
+            state = application.get_application_state()
+            if state == "Running":
+                self.masters[application.master] = True
+            else:
+                self.masters[application.master] = False
+
+    def execute(self, data):
+        self.update_running_clusters()
+        if not False in self.masters.values():
```

Comment:
maybe this "if" block should be protected by a lock.

Reply:
This was fixed by a more pythonic solution.